### ISSUE TITLE
AttributeError trying to access Redshift dialect when credentials are passed in a connection string

### ISSUE THREAD

**Issue by douglascook** [fivetran/great_expectations#4584]
**Describe the bug**
Certain expectations have been failing for me when run against Redshift, with the following exception:

```
AttributeError: module 'sqlalchemy_redshift.dialect' has no attribute 'dialect'
```

The problem appears to be caused by an import in `/great_expectations/expectations/metrics/util.py` and only occurs if the redshift dialect has not *already* been imported correctly somewhere else.

**To Reproduce**
I haven't been able to reproduce this running in a local notebook against Redshift, but am running into the issue when using the Airflow `GreatExpecationsOperator` in a DAG to run a Checkpoint.

A couple of the exceptions that I have hit are below:

```
{
  "exception_info": {
    "exception_message": "module 'sqlalchemy_redshift.dialect' has no attribute 'dialect'",
    "exception_traceback": "Traceback (most recent call last):\n  File \"/usr/local/airflow/.local/lib/python3.7/site-packages/great_expectations/execution_engine/execution_engine.py\", line 378, in resolve_metrics\n    **metric_provider_kwargs\n  File \"/usr/local/airflow/.local/lib/python3.7/site-packages/great_expectations/expectations/metrics/metric_provider.py\", line 55, in inner_func\n    return metric_fn(*args, **kwargs)\n  File \"/usr/local/airflow/.local/lib/python3.7/site-packages/great_expectations/expectations/metrics/map_metric_provider.py\", line 406, in inner_func\n    _metrics=metrics,\n  File \"/usr/local/airflow/.local/lib/python3.7/site-packages/great_expectations/expectations/metrics/column_map_metrics/column_values_unique.py\", line 49, in _sqlalchemy_window\n    if sql_engine and dialect and dialect.dialect.name == \"mysql\":\nAttributeError: module 'sqlalchemy_redshift.dialect' has no attribute 'dialect'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/usr/local/airflow/.local/lib/python3.7/site-packages/great_expectations/validator/validator.py\", line 1115, in resolve_validation_graph\n    runtime_configuration=runtime_configuration,\n  File \"/usr/local/airflow/.local/lib/python3.7/site-packages/great_expectations/validator/validator.py\", line 2017, in _resolve_metrics\n    runtime_configuration=runtime_configuration,\n  File \"/usr/local/airflow/.local/lib/python3.7/site-packages/great_expectations/execution_engine/execution_engine.py\", line 382, in resolve_metrics\n    message=str(e), failed_metrics=(metric_to_resolve,)\ngreat_expectations.exceptions.exceptions.MetricResolutionError: module 'sqlalchemy_redshift.dialect' has no attribute 'dialect'\n",
    "raised_exception": true
  },
  "expectation_config": {
    "expectation_type": "expect_column_values_to_be_unique",
    "kwargs": {
      "batch_id": "797f86579eff0f1862762a7f6ff6faba",
      "column": "partner_id"
    },
  },
},
{
  "exception_info": {
    "exception_message": "module 'sqlalchemy_redshift.dialect' has no attribute 'dialect'",
    "exception_traceback": "Traceback (most recent call last):\n  File \"/usr/local/airflow/.local/lib/python3.7/site-packages/great_expectations/execution_engine/execution_engine.py\", line 378, in resolve_metrics\n    **metric_provider_kwargs\n  File \"/usr/local/airflow/.local/lib/python3.7/site-packages/great_expectations/expectations/metrics/metric_provider.py\", line 55, in inner_func\n    return metric_fn(*args, **kwargs)\n  File \"/usr/local/airflow/.local/lib/python3.7/site-packages/great_expectations/expectations/metrics/map_metric_provider.py\", line 406, in inner_func\n    _metrics=metrics,\n  File \"/usr/local/airflow/.local/lib/python3.7/site-packages/great_expectations/expectations/metrics/column_map_metrics/column_values_match_regex.py\", line 30, in _sqlalchemy\n    f\"Regex is not supported for dialect {str(_dialect.dialect.name)}\"\nAttributeError: module 'sqlalchemy_redshift.dialect' has no attribute 'dialect'\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/usr/local/airflow/.local/lib/python3.7/site-packages/great_expectations/validator/validator.py\", line 1115, in resolve_validation_graph\n    runtime_configuration=runtime_configuration,\n  File \"/usr/local/airflow/.local/lib/python3.7/site-packages/great_expectations/validator/validator.py\", line 2017, in _resolve_metrics\n    runtime_configuration=runtime_configuration,\n  File \"/usr/local/airflow/.local/lib/python3.7/site-packages/great_expectations/execution_engine/execution_engine.py\", line 382, in resolve_metrics\n    message=str(e), failed_metrics=(metric_to_resolve,)\ngreat_expectations.exceptions.exceptions.MetricResolutionError: module 'sqlalchemy_redshift.dialect' has no attribute 'dialect'\n",
    "raised_exception": true
  },
  "expectation_config": {
    "expectation_type": "expect_column_values_to_match_regex",
    "kwargs": {
      "batch_id": "797f86579eff0f1862762a7f6ff6faba",
      "column": "name",
      "regex": "^[a-z]+$"
    },
  },
},
```

The problem appears to be caused by the import here: 

https://github.com/great-expectations/great_expectations/blob/27d2dbd8834fea58c8746f72a3c648766d434db5/great_expectations/expectations/metrics/util.py#L56 

and (for example) the subsequent attempt to access the dialect as an attribute here:

https://github.com/great-expectations/great_expectations/blob/27d2dbd8834fea58c8746f72a3c648766d434db5/great_expectations/expectations/metrics/util.py#L136

Unless the dialect module has already been imported correctly with `import sqlalchemy_redshift.dialect` or `from sqlalchemy_redshift import dialect` it is not available as an attribute on the top level package. I haven't dug into the notebooks but I imagine that this is why the expectations work when running from there.

**Environment (please complete the following information):**
 - Operating System: Linux, running on [AWS MWAA ](https://docs.aws.amazon.com/mwaa/index.html)
 - Great Expectations Version: `0.14.11`

I'm happy to have a go at trying to fix this myself and should be able to get a PR open soon assuming it doesn't turn out to be more complicated than I thought!

**Comment by douglascook** [comment]
Looking into this a bit more I found the helpful comment below:

https://github.com/great-expectations/great_expectations/blob/c555d9fd20f70df4a70bb0102fad2b57c1556ca5/great_expectations/self_check/util.py#L147-L152

It looks the error I'm running into could be the same issue, with the [sqlalchemy-redshift](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift) dialect not providing a `dialect` attribute

**Comment by douglascook** [comment]
I've managed to reproduce this using the script below. 

I've also narrowed down the problem to the use of `connection_string` in the execution engine config. Everything works correctly when the credentials are passed individually via the `credentials` key

My local environment details:

* macOS 11.6.5
* python 3.7.13
* great-expectations 0.14.11
* SQLAlchemy 1.3.24
* sqlalchemy-redshift 0.8.6

**Steps to Reproduce**

1. Create a `master.partner` table in Redshift with INT `partner_id` and VARCHAR `name` columns.
2. Set up expectations dir with the expectation config below saved as `partner.json`
3. Export env var  containing your `REDSHIFT_CONN_STRING`, starting with `redshift+psycopg2://`
4. Run the script and the expectations will fail with the expected exception

```json
{
  "expectation_suite_name": "partner",
  "expectations": [
    {
      "expectation_type": "expect_column_values_to_be_unique",
      "kwargs": {
        "column": "partner_id"
      }
    },
    {
      "expectation_type": "expect_column_values_to_match_regex",
      "kwargs": {
        "column": "name",
        "regex": "^[a-z]+$"
      }
    }
  ]
}
```

```python
import os

import yaml
from great_expectations.data_context import BaseDataContext
from great_expectations.data_context.types.base import (
    DataContextConfig,
    DatasourceConfig,
)
from great_expectations.data_context.util import instantiate_class_from_config

REDSHIFT_CONN_STRING = os.environ["REDSHIFT_CONN_STRING"]

CHECKPOINT_YML = """
name: master.partner
config_version: 1.0
class_name: Checkpoint
module_name: great_expectations.checkpoint
validations:
  - batch_request:
      datasource_name: redshift
      data_connector_name: master_tables_data_connector
      data_asset_name: master.partner
    expectation_suite_name: partner
action_list:
  - name: store_validation_result
    action:
      class_name: StoreValidationResultAction
runtime_configuration:
  result_format:
    result_format: SUMMARY
    partial_unexpected_count: 10
"""


def build_data_context():
    return BaseDataContext(
        DataContextConfig(
            config_version=3,
            plugins_directory=None,
            config_variables_file_path=None,
            datasources={
                "redshift": DatasourceConfig(
                    name="redshift",
                    class_name="Datasource",
                    execution_engine={
                        "class_name": "SqlAlchemyExecutionEngine",
                        "connection_string": REDSHIFT_CONN_STRING,
                    },
                    data_connectors={
                        "master_tables_data_connector": {
                            "class_name": "InferredAssetSqlDataConnector",
                            "module_name": "great_expectations.datasource.data_connector",
                            "include_schema_name": True,
                        }
                    },
                )
            },
            stores={
                "expectations_store": {
                    "class_name": "ExpectationsStore",
                    "store_backend": {
                        "class_name": "TupleFilesystemStoreBackend",
                        "base_directory": "/great_expectations/expectations/",
                    },
                },
                "validations_store": {
                    "class_name": "ValidationsStore",
                    "store_backend": {
                        "class_name": "TupleFilesystemStoreBackend",
                        "base_directory": "great_expectations/validations/",
                    },
                },
                "evaluation_parameter_store": {
                    "class_name": "EvaluationParameterStore"
                },
            },
            checkpoint_store_name="checkpoint_store",
            expectations_store_name="expectations_store",
            validations_store_name="validations_store",
            evaluation_parameter_store_name="evaluation_parameter_store",
            anonymous_usage_statistics={"enabled": False},
            data_docs_sites={},
        )
    )


if __name__ == "__main__":
    context = build_data_context()
    checkpoint = instantiate_class_from_config(
        config=yaml.load(CHECKPOINT_YML),
        runtime_environment={"data_context": context},
        config_defaults={"module_name": "great_expectations.checkpoint"},
    )
    checkpoint.run()
```

**Comment by talagluck** [comment]
Hi @douglascook - thank you for raising this, and for doing a lot of the legwork in tracking down the source of the issue.

It may take us some time to get to this Issue. Is this something that you would be interested in contributing? I agree that it seems like the right fix would be something similar to the one in `self_check/util.py` for BigQuery.

Please let me know if you need any guidance or support to move forward!

**Comment by douglascook** [comment]
Hi @talagluck - yeah definitely, I'd be happy to have a look into it myself at some point.

I opened PRs for another couple of issues already but those were a bit more pressing on my side, because I'd had to hack around them in our codebase to get things working without the fixes proposed in the PRs. In this case the solution that avoids using a connection string seems pretty reasonable, so isn't as urgent for us to resolve. 

TLDR I'm certainly interested in investigating more and opening a PR but will probably take a bit longer to get round to 😁

**Comment by rdodev** [comment]
Hey @douglascook we have not been able to prioritize this fix bug. Is it still relevant and are you intending to submit a contribution?

**Comment by douglascook** [comment]
Hi @rdodev, we've migrated away from Redshift so this is no longer an issue for me personally. I won't be able to work on it either unfortunately, because I no longer have access to a Redshift cluster. Thanks for the reminder anyway!

**Comment by rdodev** [comment]
Thanks for response @douglascook. Please feel free to reopen, if it becomes a pain point again.
